File: 01-Array_String/605-can_place_flowers.cpp
/* ----------------------------------------------------
   605. Can Place Flowers (24/08/25); JDG - C++ Solution
   Time complexity : O(n)
   Space complexity: O(1)
   ---------------------------------------------------- */

#include <iostream>
#include <vector>

class Solution {
public:
    bool canPlaceFlowers(std::vector<int>& flowerbed, int n) {
        int flowerbedSize = flowerbed.size();
        int plantablePlots = 0;

        for(int i = 0; i < flowerbedSize; i++) {
            int leftEmpty = ((i == 0) || (flowerbed[i-1] == 0));
            int rightEmpty = (( i == flowerbedSize - 1) || (flowerbed[i+1] == 0));

            if((flowerbed[i] == 0) && leftEmpty && rightEmpty) {
                plantablePlots++;
                flowerbed[i] = 1;
            }
        }
        return (n <= plantablePlots);
    }
};

int main() {
    std::vector<int> flowerbed = {1,0,0,0,1};
    int n = 1;
    Solution solution;

    bool result = solution.canPlaceFlowers(flowerbed, n);
    std::cout << (result ? "true" : "false") << '\n';

    return 0;
}

File: 01-Array_String/605-can_place_flowers.py
# ----------------------------------------------------
#  608. Can Place Flowers (24/08/25); JDG - Py Solution
#  Time complexity : O(n)
#  Space complexity: O(1)
#  ----------------------------------------------------

from typing import List

class Solution:
    def canPlaceFlowers(self, flowerbed: List[int], n:int) -> bool:
        flowerbedSize = len(flowerbed)
        plantablePlots = 0

        for i in range(len(flowerbed)):
            leftEmpty = ((i == 0) or (flowerbed[i-1] == 0))
            rightEmpty = ((i == (flowerbedSize - 1)) or flowerbed[i+1] == 0)

            if(i == 0 and leftEmpty and rightEmpty):
                plantablePlots+=1
                flowerbed[i] = 1

        return (n <= plantablePlots)

if __name__ == "__main__":
    flowerbed = [1,0,0,0,1]
    n = 1
    solution = Solution()

    print(solution.canPlaceFlowers(flowerbed, n))
